File: yasmin_ros/tests/python/test_yasmin_ros.py
# Copyright (C) 2023 Miguel Ángel González Santamarta

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


import time
import unittest
from threading import Thread

from yasmin_ros.ros_communications_cache import ROSCommunicationsCache
from yasmin_ros import ActionState, ServiceState, MonitorState, PublisherState
from yasmin_ros.basic_outcomes import SUCCEED, CANCEL, ABORT, TIMEOUT

from example_interfaces.action import Fibonacci
from example_interfaces.srv import AddTwoInts
from std_msgs.msg import String

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.action import CancelResponse, GoalResponse


class AuxNode(Node):
    def __init__(self):
        super().__init__("test_node")

        self.action_server = ActionServer(
            self,
            Fibonacci,
            "test",
            goal_callback=self.goal_callback,
            handle_accepted_callback=self.handle_accepted_callback,
            execute_callback=self.execute_action,
            cancel_callback=self.cancel_action,
            callback_group=ReentrantCallbackGroup(),
        )

        self.service_server = self.create_service(
            AddTwoInts, "test", self.execute_service
        )

        self.pub = self.create_publisher(String, "test", 10)
        self.timer = self.create_timer(1, self.publish_msgs)

    def goal_callback(self, goal_request) -> int:
        return GoalResponse.ACCEPT

    def handle_accepted_callback(self, goal_handle) -> None:
        goal_handle.execute()

    def execute_action(self, goal_handle):
        result = Fibonacci.Result()

        request = goal_handle.request

        if request.order < 0:
            goal_handle.abort()

        else:
            time.sleep(3)

            if goal_handle.is_cancel_requested:
                goal_handle.canceled()
            else:
                goal_handle.succeed()

        return result

    def cancel_action(self, goal_handle) -> int:
        return CancelResponse.ACCEPT

    def execute_service(self, request, response):
        time.sleep(2)
        response.sum = request.a + request.b
        return response

    def publish_msgs(self) -> None:
        msg = String()
        msg.data = "data"
        self.pub.publish(msg)


class TestYasminRos(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        rclpy.init()

        cls.aux_node = AuxNode()
        cls.executor = MultiThreadedExecutor()
        cls.executor.add_node(cls.aux_node)

        cls.spin_thread = Thread(target=cls.executor.spin)
        cls.spin_thread.start()

    @classmethod
    def tearDownClass(cls):
        rclpy.shutdown()

    def test_action_succeed(self):

        def create_goal_cb(blackboard):
            goal = Fibonacci.Goal()
            goal.order = 0
            return goal

        state = ActionState(Fibonacci, "test", create_goal_cb)
        self.assertEqual(SUCCEED, state())

    def test_action_client_cache(self):
        ROSCommunicationsCache.clear_all()
        self.assertEqual(0, ROSCommunicationsCache.get_action_clients_count())

        def create_goal_cb(blackboard):
            goal = Fibonacci.Goal()
            goal.order = 0
            return goal

        state1 = ActionState(Fibonacci, "test", create_goal_cb)
        self.assertEqual(1, ROSCommunicationsCache.get_action_clients_count())

        state2 = ActionState(Fibonacci, "test", create_goal_cb)
        self.assertEqual(1, ROSCommunicationsCache.get_action_clients_count())

        state3 = ActionState(Fibonacci, "test2", create_goal_cb)
        self.assertEqual(2, ROSCommunicationsCache.get_action_clients_count())

    def test_action_result_handler(self):

        def create_goal_cb(blackboard):
            goal = Fibonacci.Goal()
            goal.order = 0
            return goal

        def result_handler(blackboard, result):
            return "new_outcome"

        state = ActionState(
            Fibonacci, "test", create_goal_cb, ["new_outcome"], result_handler
        )
        self.assertEqual("new_outcome", state())

    def test_action_cancel(self):

        def create_goal_cb(blackboard):
            goal = Fibonacci.Goal()
            goal.order = 20
            return goal

        def cancel_state(state, seconds):
            time.sleep(seconds)
            state.cancel_state()

        state = ActionState(Fibonacci, "test", create_goal_cb)
        thread = Thread(
            target=cancel_state,
            args=(
                state,
                1,
            ),
        )
        thread.start()
        self.assertEqual(CANCEL, state())
        thread.join()

    def test_action_abort(self):

        def create_goal_cb(blackboard):
            goal = Fibonacci.Goal()
            goal.order = -1
            return goal

        state = ActionState(Fibonacci, "test", create_goal_cb)
        self.assertEqual(ABORT, state())

    def test_action_retry_wait_timeout(self):
        def create_goal_cb(blackboard):
            goal = Fibonacci.Goal()
            goal.order = 3
            return goal

        retries = 3

        ## Capture the logs
        with self.assertLogs("root", level="WARNING") as captured:
            state = ActionState(
                Fibonacci,
                "test1",
                create_goal_cb,
                wait_timeout=0.1,
                maximum_retry=retries,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            (retries * 2) + 1,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )

    def test_action_retry_response_timeout(self):
        def create_goal_cb(blackboard):
            goal = Fibonacci.Goal()
            goal.order = 3
            return goal

        retries = 3

        ## Capture the logs
        with self.assertLogs("root", level="WARNING") as captured:
            state = ActionState(
                Fibonacci,
                "test",
                create_goal_cb,
                response_timeout=0.1,
                maximum_retry=retries,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            (retries * 2) + 1,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )

    def test_service(self):

        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        state = ServiceState(AddTwoInts, "test", create_request_cb)
        self.assertEqual(SUCCEED, state())

    def test_service_client_cache(self):
        ROSCommunicationsCache.clear_all()
        self.assertEqual(0, ROSCommunicationsCache.get_service_clients_count())

        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        state1 = ServiceState(AddTwoInts, "test", create_request_cb)
        self.assertEqual(1, ROSCommunicationsCache.get_service_clients_count())

        state2 = ServiceState(AddTwoInts, "test", create_request_cb)
        self.assertEqual(1, ROSCommunicationsCache.get_service_clients_count())

        state3 = ServiceState(AddTwoInts, "test2", create_request_cb)
        self.assertEqual(2, ROSCommunicationsCache.get_service_clients_count())

    def test_service_response_handler(self):

        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        def response_handler(blackboard, response):
            return "new_outcome"

        state = ServiceState(
            AddTwoInts, "test", create_request_cb, ["new_outcome"], response_handler
        )
        self.assertEqual("new_outcome", state())

    def test_service_retry_wait_timeout(self):
        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        retries = 3

        ## Capture the logs
        with self.assertLogs("root", level="WARNING") as captured:
            state = ServiceState(
                AddTwoInts,
                "test_retry",
                create_request_cb,
                maximum_retry=retries,
                wait_timeout=0.1,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            (retries * 2) + 1,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )

    def test_service_retry_response_timeout(self):
        def create_request_cb(blackboard):
            request = AddTwoInts.Request()
            request.a = 2
            request.b = 3
            return request

        retries = 3

        ## Capture the logs
        with self.assertLogs("root", level="WARNING") as captured:
            state = ServiceState(
                AddTwoInts,
                "test",
                create_request_cb,
                maximum_retry=retries,
                response_timeout=0.1,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            (retries * 2) + 1,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )

    def test_monitor_timeout(self):

        def monitor_handler(blackboard, msg):
            return SUCCEED

        state = MonitorState(
            String, "test1", [SUCCEED], monitor_handler=monitor_handler, timeout=2
        )
        self.assertEqual(TIMEOUT, state())

    def test_monitor_cache(self):
        ROSCommunicationsCache.clear_all()
        self.assertEqual(0, ROSCommunicationsCache.get_subscribers_count())

        def monitor_handler(blackboard, msg):
            return SUCCEED

        state1 = MonitorState(String, "test", [SUCCEED], monitor_handler)
        self.assertEqual(1, ROSCommunicationsCache.get_subscribers_count())

        state2 = MonitorState(String, "test", [SUCCEED], monitor_handler)
        self.assertEqual(1, ROSCommunicationsCache.get_subscribers_count())

        state3 = MonitorState(String, "test2", [SUCCEED], monitor_handler)
        self.assertEqual(2, ROSCommunicationsCache.get_subscribers_count())

    def test_monitor_retry_response_timeout(self):
        def monitor_handler(blackboard, msg):
            return SUCCEED

        retries = 3

        ## Capture the logs
        with self.assertLogs("root", level="WARNING") as captured:
            state = MonitorState(
                String,
                "test1",
                [SUCCEED],
                monitor_handler=monitor_handler,
                timeout=0.1,
                maximum_retry=retries,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            (retries * 2) + 1,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )

    def test_publisher_cache(self):
        ROSCommunicationsCache.clear_all()
        self.assertEqual(0, ROSCommunicationsCache.get_publishers_count())

        def create_msg_handler(blackboard):
            msg = String()
            msg.data = "data"
            return msg

        state1 = PublisherState(String, "test", create_msg_handler)
        self.assertEqual(1, ROSCommunicationsCache.get_publishers_count())

        state2 = PublisherState(String, "test", create_msg_handler)
        self.assertEqual(1, ROSCommunicationsCache.get_publishers_count())

        state3 = PublisherState(String, "test2", create_msg_handler)
        self.assertEqual(2, ROSCommunicationsCache.get_publishers_count())

File: yasmin_ros/test/test_yasmin_ros.cpp
// Copyright (C) 2025 Miguel Ángel González Santamarta
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <chrono>
#include <gtest/gtest.h>
#include <memory>
#include <thread>

#include "example_interfaces/action/fibonacci.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"
#include "std_msgs/msg/string.hpp"

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

#include "yasmin_ros/action_state.hpp"
#include "yasmin_ros/basic_outcomes.hpp"
#include "yasmin_ros/monitor_state.hpp"
#include "yasmin_ros/publisher_state.hpp"
#include "yasmin_ros/ros_clients_cache.hpp"
#include "yasmin_ros/ros_logs.hpp"
#include "yasmin_ros/service_state.hpp"
#include "yasmin_ros/yasmin_node.hpp"

using namespace yasmin_ros;
using namespace yasmin_ros::basic_outcomes;
using namespace std::chrono_literals;

class AuxNode : public rclcpp::Node {
public:
  using Fibonacci = example_interfaces::action::Fibonacci;
  using GoalHandleFibonacci = rclcpp_action::ServerGoalHandle<Fibonacci>;
  using AddTwoInts = example_interfaces::srv::AddTwoInts;

  AuxNode() : Node("test_node") {
    // Action server
    action_server_ = rclcpp_action::create_server<Fibonacci>(
        this, "test",
        std::bind(&AuxNode::handle_goal, this, std::placeholders::_1,
                  std::placeholders::_2),
        std::bind(&AuxNode::handle_cancel, this, std::placeholders::_1),
        std::bind(&AuxNode::handle_accepted, this, std::placeholders::_1));

    // Service server
    service_server_ = this->create_service<AddTwoInts>(
        "test", std::bind(&AuxNode::handle_service, this, std::placeholders::_1,
                          std::placeholders::_2));

    // Publisher
    publisher_ = this->create_publisher<std_msgs::msg::String>("test", 10);
    timer_ =
        this->create_wall_timer(1s, std::bind(&AuxNode::publish_message, this));
  }

private:
  rclcpp_action::Server<Fibonacci>::SharedPtr action_server_;
  rclcpp::Service<AddTwoInts>::SharedPtr service_server_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;

  rclcpp_action::GoalResponse
  handle_goal(const rclcpp_action::GoalUUID &uuid,
              std::shared_ptr<const Fibonacci::Goal> goal) {
    (void)uuid;
    (void)goal;
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  rclcpp_action::CancelResponse
  handle_cancel(const std::shared_ptr<GoalHandleFibonacci> goal_handle) {
    (void)goal_handle;
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  void handle_accepted(const std::shared_ptr<GoalHandleFibonacci> goal_handle) {
    std::thread{
        std::bind(&AuxNode::execute_action, this, std::placeholders::_1),
        goal_handle}
        .detach();
  }

  void execute_action(const std::shared_ptr<GoalHandleFibonacci> goal_handle) {
    const auto goal = goal_handle->get_goal();
    auto result = std::make_shared<Fibonacci::Result>();

    if (goal->order < 0) {
      goal_handle->abort(result);
    } else {
      std::this_thread::sleep_for(5s);
      if (goal_handle->is_canceling()) {
        goal_handle->canceled(result);
      } else {
        goal_handle->succeed(result);
      }
    }
  }

  void handle_service(const std::shared_ptr<AddTwoInts::Request> request,
                      std::shared_ptr<AddTwoInts::Response> response) {
    std::this_thread::sleep_for(5s);
    response->sum = request->a + request->b;
  }

  void publish_message() {
    if (rclcpp::ok()) {
      auto message = std_msgs::msg::String();
      message.data = "data";
      publisher_->publish(message);
    }
  }
};

class TestYasminRos : public ::testing::Test {
protected:
  static std::shared_ptr<AuxNode> aux_node;
  static std::shared_ptr<rclcpp::executors::MultiThreadedExecutor> executor;
  static std::thread spin_thread;

  static void SetUpTestSuite() {
    rclcpp::init(0, nullptr);
    aux_node = std::make_shared<AuxNode>();
    executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
    executor->add_node(aux_node);
    spin_thread = std::thread([&]() { executor->spin(); });
  }

  static void TearDownTestSuite() {
    executor->cancel();
    if (spin_thread.joinable()) {
      spin_thread.join();
    }
    ROSClientsCache::clear_all();
    aux_node.reset();
    executor.reset();
    rclcpp::shutdown();
  }
};

std::shared_ptr<AuxNode> TestYasminRos::aux_node = nullptr;
std::shared_ptr<rclcpp::executors::MultiThreadedExecutor>
    TestYasminRos::executor = nullptr;
std::thread TestYasminRos::spin_thread;

TEST_F(TestYasminRos, TestActionSucceed) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 0;
            return goal;
          });

  EXPECT_EQ((*state)(blackboard), std::string(SUCCEED));
}

TEST_F(TestYasminRos, TestActionClientCache) {
  ROSClientsCache::clear_all();
  EXPECT_EQ(ROSClientsCache::get_action_clients_count(), 0);

  auto state1 =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 0;
            return goal;
          });
  EXPECT_EQ(ROSClientsCache::get_action_clients_count(), 1);

  auto state2 =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 0;
            return goal;
          });
  EXPECT_EQ(ROSClientsCache::get_action_clients_count(), 1);

  auto state3 =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test2",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 0;
            return goal;
          });
  EXPECT_EQ(ROSClientsCache::get_action_clients_count(), 2);
}

TEST_F(TestYasminRos, TestActionResultHandler) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 0;
            return goal;
          },
          std::set<std::string>{"new_outcome"},
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
             std::shared_ptr<example_interfaces::action::Fibonacci::Result>
                 result) { return "new_outcome"; });

  EXPECT_EQ((*state)(blackboard), "new_outcome");
}

TEST_F(TestYasminRos, TestActionCancel) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 20;
            return goal;
          });

  std::thread cancel_thread([&state]() {
    std::this_thread::sleep_for(1s);
    state->cancel_state();
  });

  EXPECT_EQ((*state)(blackboard), std::string(CANCEL));
  cancel_thread.join();
}

TEST_F(TestYasminRos, TestActionAbort) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = -1;
            return goal;
          });

  EXPECT_EQ((*state)(blackboard), std::string(ABORT));
}

TEST_F(TestYasminRos, TestActionRetryWaitTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test1",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 3;
            return goal;
          },
          std::set<std::string>{}, 1, -1, 3); // wait_timeout=1, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestYasminRos, TestActionRetryResponseTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ActionState<example_interfaces::action::Fibonacci>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto goal = example_interfaces::action::Fibonacci::Goal();
            goal.order = 3;
            return goal;
          },
          std::set<std::string>{}, -1, 1,
          3); // response_timeout=1, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestYasminRos, TestService) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          -1, -1, 3); // wait_timeout, response_timeout, maximum_retry

  EXPECT_EQ((*state)(blackboard), std::string(SUCCEED));
}

TEST_F(TestYasminRos, TestServiceClientCache) {
  ROSClientsCache::clear_all();
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 0);

  auto state1 = std::make_shared<
      ServiceState<example_interfaces::srv::AddTwoInts>>(
      "test", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto request =
            std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = 2;
        request->b = 3;
        return request;
      });
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 1);

  auto state2 = std::make_shared<
      ServiceState<example_interfaces::srv::AddTwoInts>>(
      "test", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto request =
            std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = 2;
        request->b = 3;
        return request;
      });
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 1);

  auto state3 = std::make_shared<
      ServiceState<example_interfaces::srv::AddTwoInts>>(
      "test2", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto request =
            std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = 2;
        request->b = 3;
        return request;
      });
  EXPECT_EQ(ROSClientsCache::get_service_clients_count(), 2);
}

TEST_F(TestYasminRos, TestServiceResponseHandler) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          std::set<std::string>{"new_outcome"},
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
             std::shared_ptr<example_interfaces::srv::AddTwoInts::Response>
                 response) { return "new_outcome"; });

  EXPECT_EQ((*state)(blackboard), "new_outcome");
}

TEST_F(TestYasminRos, TestServiceRetryWaitTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test_retry",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          1, -1, 3); // wait_timeout=1, response_timeout=-1, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestYasminRos, TestServiceRetryResponseTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state =
      std::make_shared<ServiceState<example_interfaces::srv::AddTwoInts>>(
          "test",
          [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
            auto request = std::make_shared<
                example_interfaces::srv::AddTwoInts::Request>();
            request->a = 2;
            request->b = 3;
            return request;
          },
          -1, 1, 3); // wait_timeout=-1, response_timeout=1, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestYasminRos, TestMonitorTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state = std::make_shared<MonitorState<std_msgs::msg::String>>(
      "test1", std::set<std::string>{std::string(SUCCEED)},
      [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
         std::shared_ptr<std_msgs::msg::String> msg) {
        return std::string(SUCCEED);
      },
      rclcpp::QoS(10), 10, 2);

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestYasminRos, TestMonitorRetryTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state = std::make_shared<MonitorState<std_msgs::msg::String>>(
      "test1", std::set<std::string>{std::string(SUCCEED)},
      [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
         std::shared_ptr<std_msgs::msg::String> msg) {
        return std::string(SUCCEED);
      },
      rclcpp::QoS(10), 10, 2, 3); // timeout=2, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestYasminRos, TestPublisher) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state = std::make_shared<PublisherState<std_msgs::msg::String>>(
      "test", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto msg = std_msgs::msg::String();
        msg.data = "data";
        return msg;
      });

  EXPECT_EQ((*state)(blackboard), std::string(SUCCEED));
}

TEST_F(TestYasminRos, TestPublisherCache) {
  ROSClientsCache::clear_all();
  EXPECT_EQ(ROSClientsCache::get_publishers_count(), 0);

  auto state1 = std::make_shared<PublisherState<std_msgs::msg::String>>(
      "test", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto msg = std_msgs::msg::String();
        msg.data = "data";
        return msg;
      });
  EXPECT_EQ(ROSClientsCache::get_publishers_count(), 1);

  auto state2 = std::make_shared<PublisherState<std_msgs::msg::String>>(
      "test", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto msg = std_msgs::msg::String();
        msg.data = "data";
        return msg;
      });
  EXPECT_EQ(ROSClientsCache::get_publishers_count(), 1);

  auto state3 = std::make_shared<PublisherState<std_msgs::msg::String>>(
      "test2", [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard) {
        auto msg = std_msgs::msg::String();
        msg.data = "data";
        return msg;
      });
  EXPECT_EQ(ROSClientsCache::get_publishers_count(), 2);
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
